File: Math/leetcode_50.cpp

// Problem link ===>> https://leetcode.com/problems/powx-n/description/

#include <bits/stdc++.h>
#define ll long long
using namespace std;

class Solution
{
public:
    double myPow(double x, int n)
    {

        if (n < 0)
        {
            x = 1 / x;
        }

        long num = labs(n);

        double pow = 1;

        while (num)
        { // equivalent to while(num != 0)
            if (num & 1)
            { // equivalent to if((num & 1) != 0)
                pow *= x;
            }

            x *= x;
            num >>= 1;
        }

        return pow;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);
    
    ll t;
    cin >> t;
    while (t--)
    {
        ll n;
        cin >> n;

        
    }
    return 0;
}

File: Math/leetcode_50.py

# Problem link ===>>  https://leetcode.com/problems/powx-n/description/
# submission link == =>> https://leetcode.com/problems/powx-n/submissions/1497362495/


class List:
    def __init__(self):
        self.data = []

    def append(self, item):
        self.data.append(item)


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n<0:
            x = 1/x
            n = -n
        
        pow_result = 1.0
        num = abs(n) 
        
        while num:
            if num & 1:  # If the least significant bit is set
                pow_result *= x
            x *= x  # Square the base
            num >>= 1  # Right shift num by 1 bit
            
        return pow_result
        
        


if __name__ == "__main__":
    solution = Solution()
